### PR TITLE
add ansible-vault secrets support;

### PR DESCRIPTION
**Pull request by f100024**
Please take a look. Refers to the issue https://github.com/umputun/spot/issues/120

### CODE REVIEW

File: pkg/secrets/ansible_vault.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	}
+	// Decrypt ansible-vault
+	decryptedVault, err := vault.DecryptFile(vault_path, secret)
+	if err != nil {
+		return nil, fmt.Errorf("error decrypting file: %s", vault_path)
+	}
+	log.Printf("[INFO] ansible vault file decrypted")
+
+	// Unmarshal decrypted data
+	m := make(map[interface{}]interface{})
```
Comment (umputun) [inline pkg/secrets/ansible_vault.go:38]: I don't get this part. is key can be non-string? Same question about value, i.e. why interface{}?

Another question - why the content parsed as yaml? Is it stored as yaml inside the vault? Docs says "You can encrypt single values inside a YAML file using ..." but I think by yaml they mean the playbook. Your code seems to expect key:value yaml inside, and if this is the case we need some clarification with comments explaining how this stored inside with some links to docs.

**Review by umputun — CHANGES_REQUESTED** [review]
Thx. I have several suggestions, but the biggest problem is probably my lack of understanding how all of this supposed to work. I have not used much of ansible vault outside of encrypting/deceypting variables like described here https://docs.ansible.com/ansible/latest/vault_guide/vault_encrypting_content.html#encrypting-individual-variables-with-ansible-vault

Your implementation expects ansible vault file, which is yml dictionary inside. How those even created, can you point be to the corresponding docs?

**Comment by umputun** [comment]
And one more thing - it is missing any changes in README.md describing all of this

File: pkg/secrets/ansible_vault.go (hunk 1 above)
Comment (f100024) [reply pkg/secrets/ansible_vault.go]: >  Another question - why is the content parsed as yaml?

`YAML` format has been chosen as common for variables in `ansible`.

> Is it stored as yaml inside the vault?

Yes.

> Docs says, "You can encrypt single values inside a YAML file using ..." but I think by yaml they mean the playbook.

Yes, you can use as single encrypted values created with parameter `encrypt-string` [link](https://docs.ansible.com/ansible/latest/vault_guide/vault_encrypting_content.html#creating-encrypted-variables), and this feature has not been implemented. Also you can use full encrypted files [link](https://docs.ansible.com/ansible/latest/vault_guide/vault_encrypting_content.html#encrypting-files-with-ansible-vault).

**Comment by f100024** [comment]
> Your implementation expects ansible vault file, which is yml dictionary inside. How were those even created? 
> Can you point be to the corresponding docs?
For saving data in ansilble vault user follow next steps:
1. Locally create `YAML` file with secrets in next format:
```
---
key1: value
key2: value
key3: value
```
2. Encrypt created file using ansible vault and [manual](https://docs.ansible.com/ansible/latest/cli/ansible-vault.html#encrypt)
Example:
```
$ ansible-vault encrypt [path to yaml file with secrets]
and enter as secret phrase word `password`
```
`ansible-vault` will create fully encrypted file which may be save in repository:
```
$ANSIBLE_VAULT;1.1;AES256
32353064613336636531383331363433323439383037306130316539363531333932323230373461
3362366235623165313931643430333437623230356265660a303631646636616539373264643865
37346631613563663939316135623261636334376431366333663330616138393038373330663334
6530306437306239650a346133633933633862623765383365376662643235386165356231373136
65663032326164363836363439393966346237306337663361623430386562326463396466313563
3830613730663039373937633433636563393535613763376164
```

Note: current implementation does not support separate encrypted values.

Corresponding [docs](https://docs.ansible.com/ansible/latest/vault_guide/vault_managing_passwords.html).

File: cmd/spot/main.go [hunk 2]
```diff
@@ -79,6 +79,11 @@ type SecretsProvider struct {
 		AccessKey string `long:"access-key" env:"ACCESS_KEY" description:"aws access key"`
 		SecretKey string `long:"secret-key" env:"SECRET_KEY" description:"aws secret key"`
 	} `group:"aws" namespace:"aws" env-namespace:"AWS"`
+
+	AnsibleVault struct {
+		VaultPath   string `long:"vault_path" env:"VAULT_PATH" description:"path to the ansible-vault file"`
```
[the remark this answers is not in the document]
Comment (f100024) [reply cmd/spot/main.go]: Done in https://github.com/f100024/spot/pull/1/commits/41b34f7bc8db04348f0049291242adc7218c232f

File: pkg/secrets/ansible_vault.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+// Get decrypted data from ansible-vault file
+func (p *AnsibleVaultProvider) Get(key string) (string, error) {
+	if key_value, ok := p.data[key]; ok {
+		return fmt.Sprintf("%v", key_value), nil
+	}
+	return "", fmt.Errorf("not found key: %v", key)
+}
+
+func GetVaultSecret(secret string) (string, error) {
```
[the remark this answers is not in the document]
Comment (f100024) [reply pkg/secrets/ansible_vault.go]: Done in https://github.com/f100024/spot/pull/1/commits/d8b08e3e25f764835e4aa692e83de85ede94c1c8

File: pkg/secrets/ansible_vault_test.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package secrets
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestAnsibleVaultProvider_Get(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (f100024) [reply pkg/secrets/ansible_vault_test.go:10]: Added tests in https://github.com/f100024/spot/pull/1/commits/83cfb218b7177ed9349da01a5813bed078c2a651

File: pkg/secrets/ansible_vault.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	"syscall"
+
+	vault "github.com/sosedoff/ansible-vault-go"
+	term "golang.org/x/term"
+	yaml "gopkg.in/yaml.v3"
+)
+
+// AnsibleVaultProvider is a provider for ansible-vault files
+type AnsibleVaultProvider struct {
+	vault_file_path string
```
[the remark this answers is not in the document]
Comment (f100024) [reply pkg/secrets/ansible_vault.go]: Done in https://github.com/f100024/spot/pull/1/commits/d347bae6bb9626a10ba30027a77a5d7aec638728

**Comment by f100024** [comment]
> And one more thing - it is missing any changes in README.md describing all of this

Updated `README.md`  in https://github.com/f100024/spot/pull/1/commits/5987df6be916adf745e0ddbf036fd99de3d0f569

**Comment by umputun** [comment]
LGTM. Pls, address the linter's warning and rebase/squash for the final merge. I think making it two commits (the code/readme change in the firs, and the vendor change in the second) will do

**Pull request merged**